pkg/providers/instance/instance.go

```diff
@@ -360,32 +407,101 @@ func (p *DefaultProvider) getOverrides(instanceTypes []*cloudprovider.InstanceTy
-// must be explicitly included in capacity type requirements.
+// getCapacityType selects the capacity type based on the flexibility of the NodeClaim and the available offerings.
+// Prioritization is as follows: reserved, spot, on-demand.
 func (p *DefaultProvider) getCapacityType(nodeClaim *karpv1.NodeClaim, instanceTypes []*cloudprovider.InstanceType) string {
-	requirements := scheduling.NewNodeSelectorRequirementsWithMinValues(nodeClaim.Spec.Requirements...)
-	if requirements.Get(karpv1.CapacityTypeLabelKey).Has(karpv1.CapacityTypeSpot) {
-		requirements[karpv1.CapacityTypeLabelKey] = scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, karpv1.CapacityTypeSpot)
-		for _, instanceType := range instanceTypes {
-			for _, offering := range instanceType.Offerings.Available() {
-				if requirements.Compatible(offering.Requirements, scheduling.AllowUndefinedWellKnownLabels) == nil {
-					return karpv1.CapacityTypeSpot
-				}
+	for _, capacityType := range []string{karpv1.CapacityTypeReserved, karpv1.CapacityTypeSpot} {
+		requirements := scheduling.NewNodeSelectorRequirementsWithMinValues(nodeClaim.Spec.Requirements...)
+		if !requirements.Get(karpv1.CapacityTypeLabelKey).Has(capacityType) {
+			continue
+		}
+		requirements[karpv1.CapacityTypeLabelKey] = scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, capacityType)
+		for _, it := range instanceTypes {
+			if len(it.Offerings.Available().Compatible(requirements)) != 0 {
+				return capacityType
 			}
 		}
 	}
 	return karpv1.CapacityTypeOnDemand
 }
 
+// filterReservedInstanceTypes is used to filter the provided set of instance types to only include those with
+// available reserved offerings if the nodeclaim is compatible. If there are no available reserved offerings, no
+// filtering is applied.
+func (*DefaultProvider) filterReservedInstanceTypes(nodeClaimRequirements scheduling.Requirements, instanceTypes []*cloudprovider.InstanceType) []*cloudprovider.InstanceType {
+	nodeClaimRequirements[karpv1.CapacityTypeLabelKey] = scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, karpv1.CapacityTypeReserved)
+	var reservedInstanceTypes []*cloudprovider.InstanceType
+	for _, it := range instanceTypes {
+		// We only want to include a single offering per pool (instance type / AZ combo). This is due to a limitation in the
+		// CreateFleet API, which limits calls to specifying a single override per pool. We'll choose to launch into the pool
+		// with the most capacity.
+		zonalOfferings := map[string]*cloudprovider.Offering{}
+		for _, o := range it.Offerings.Available().Compatible(nodeClaimRequirements) {
+			if current, ok := zonalOfferings[o.Zone()]; !ok || o.ReservationCapacity > current.ReservationCapacity {
+				zonalOfferings[o.Zone()] = o
+			}
+		}
+		if len(zonalOfferings) == 0 {
+			continue
+		}
+		it.Offerings = lo.Values(zonalOfferings)
```

Comment:
This mutation of our instance type data makes me nervous -- we don't mutate it anywhere else AFAIK -- are we 100% sure that this is always safe and are there better ways to do this that don't require instance type mutation?

Reply:
Yeah, I've updated the `GetInstanceTypes()` call to return a slightly deeper shallow copy. So we can mutate the slice, we just can't mutate any of the elements of the slice. We could be a little more defensive about things and track this separately, but the deeper copy was necessary since we're no longer caching all offerings anyway.

Reply:
I think we mentioned we wanted this same comment elsewhere around the place where the offerings are constructed -- this was Friday, though, so I'm trying to remember where else we wanted it :)